datafusion/core/src/datasource/statistics.rs

```diff
@@ -57,83 +44,36 @@ pub async fn get_statistics_with_limit(
-            min_values[index] = file_column.min_value;
-        }
+        stats_agg.update(&file_stats, &file_schema)?;
 
         // If the number of rows exceeds the limit, we can stop processing
         // files. This only applies when we know the number of rows. It also
         // currently ignores tables that have no statistics regarding the
         // number of rows.
-        let conservative_num_rows = match num_rows {
-            Precision::Exact(nr) => nr,
+        let conservative_num_rows = match stats_agg.num_rows() {
+            Precision::Exact(nr) => *nr,
             _ => usize::MIN,
         };
-        if conservative_num_rows <= limit.unwrap_or(usize::MAX) {
+
+        if conservative_num_rows <= limit {
             while let Some(current) = all_files.next().await {
                 let (file, file_stats) = current?;
                 result_files.push(file);
-
-                // We accumulate the number of rows, total byte size and null
-                // counts across all the files in question. If any file does not
-                // provide any information or provides an inexact value, we demote
-                // the statistic precision to inexact.
-                num_rows = add_row_stats(file_stats.num_rows, num_rows);
-
-                total_byte_size =
-                    add_row_stats(file_stats.total_byte_size, total_byte_size);
-
-                (null_counts, max_values, min_values) = multiunzip(
-                    izip!(
-                        file_stats.column_statistics.into_iter(),
-                        null_counts.into_iter(),
-                        max_values.into_iter(),
-                        min_values.into_iter()
-                    )
-                    .map(
-                        |(
-                            ColumnStatistics {
-                                null_count: file_nc,
-                                max_value: file_max,
-                                min_value: file_min,
-                                distinct_count: _,
-                            },
-                            null_count,
-                            max_value,
-                            min_value,
-                        )| {
-                            (
-                                add_row_stats(file_nc, null_count),
-                                set_max_if_greater(file_max, max_value),
-                                set_min_if_lesser(file_min, min_value),
-                            )
-                        },
-                    ),
-                );
+                stats_agg.update(&file_stats, &file_schema)?;
 
                 // If the number of rows exceeds the limit, we can stop processing
-                // files. This only applies when we know the number of rows. It also
-                // currently ignores tables that have no statistics regarding the
-                // number of rows.
-                if num_rows.get_value().unwrap_or(&usize::MIN)
```

Comment:
I think this check may be incorrect in the sense that even if the file statistics are not precise, it may stop reading files early